### ISSUE TITLE
Trailing spaces in comments and line contination characters

### ISSUE THREAD

**Issue by jgwinner** [adrienverge/yamllint#445]
Two feature requests for trailing spaces: 

Add 'trailing-space-backslash' - which triggers on a "\  " character at EOL. 
That one is killer; bash will ignore a \ as a continuation character if there is whitespace behind it. This is what "trailing-spaces" was born for! 

However .... 

Second request:

Break up trailing-spaces to 2-3 different errors.

Change the "trailing-spaces" error to work on NON comment lines, and a separate error for 'trailing-spaces-comments' (or trailing-comment-spaces or whatever) error.

I have to admit, I have muscle memory, when I type a period or comma, I hit the space bar. This is causing > 90% of my 'trailing space' errors to be inside comments. ("Virtual Space" in my IDE may have also changed, it used to trim it).

Add a 'blank-line' error. YAML allows blank lines, but the "trailing-space" error comes up on blank lines. When you hit "Enter" in the IDE, it will automatically add spaces to match the previous line's indentation. This causes a failure to post code. 

I don't see a trailing space in a comment as an error, but a trailing space at the end of a bash continuation line is a killer, and I have no way to disambiguate them. It would be a real quality of life issue if trailing spaces at the end of non-comment lines and blank lines could be ignored.

**Comment by adrienverge** [comment]
Hello @jgwinner, thanks for sharing your thoughts.

I'm not sure I understood everything, so please correct if I'm wrong: you would prefer yamllint to report `trailing-spaces` problems differently, depending whether they're:
\- after actual code,
\- after a comment,
\- inside a blank line.

If my understanding is wrong, could you rephrase with precise examples of what you would expect?

To be honest you're the first person I see preferring to keep trailing spaces in comments and blank lines :slightly_smiling_face: Usually people use a linter to remove every non-needed characters (e.g. trailing spaces, trailing empty lines...) May I ask why you don't prefer these spaces to be removed (after they've been added by your muscle memory)?

**Comment by jgwinner** [comment]
Thanks for taking a look @adrienverge and prompt response. Yes, your understanding is correct. THere are actually two different ways to go and I think my explanation was slightly confusing.

> May I ask why you don't prefer these spaces to be removed (after they've been added by your muscle memory)?

Sure. I'd love it if the linter automatically removed them - the issue is that I'm actually using the yamllint in a github action though, and there's no opportunity for the modified code to be subsequently checked in without some GitHub actions shenanigans involving tokens.

Therefore, the checked-in code would still have the spaces in the wrong places, and thus executed by GitHub actions when they trigger.

My thought was that I could make spaces after actual code be enabled, and change to after a comment or inside a blank line to be at most warnings, or ignored, on the GitHub Action validator. 

I guess we could require developers to run yamllint before checking in code, and then the final GH Action check would blow chunks if blank lines were in the YAML. That could work, although it adds a manual step, at least for me, as I don't see a Visual Studio integration that runs yamllint. Still, it's a lot better than doing search and replace which is what I did to clean up the files.

I've found that some IDE's will automatically insert spaces in a blank line to match the indentation; even doing a regex search and replace, if you just arrow through the line the spaces get added oddly.

**Comment by jgwinner** [comment]
TL;DR:

> 
> I'm not sure I understood everything, so please correct if I'm wrong: you would prefer yamllint to report trailing-spaces problems differently, depending whether they're:
> - after actual code,
> - after a comment,
> - inside a blank line.

Correct. 3 checks, each could be setup as warning or enabled at will through the YAMLLINT config file.

**Comment by adrienverge** [comment]
OK @jgwinner, thanks it's clear to me :+1: 

I'm not completely opposed to such a change, but I'm not sure it will useful to many people. I believe most users want *their IDE to remove trailing spaces*, instead of asking their linter to tolerate some special cases. I understand you use Visual Studio, are you sure you can't find an option or plugin to do this automatically? For example I'm using Vim, and it does automatically insert spaces in a blank line to match the indentation, but removes them afterwards if the lines is blank.

**Comment by jgwinner** [comment]
It's weird to me, Visual Studio (not code) used to do this automatically. Not sure when it regressed, I just recently upgraded to 2022.

For anyone else stumbling on this:

I found one extension, "CodeMaid" but it required some fiddling (yml/yaml isn't built in, but 'text' is, and that works). There's a strange side effect where the cursor moves to the start of the next line on save, but I can live with that.

Another, "Trim line ends" seems to be more targeted.

So, those will work and I can use the indiscriminate error. (AWS's own YAML fails ironically)

**Comment by Mumeii** [comment]
Hi @adrienverge 👋 

I'm here to upvote the request of @jgwinner, at least in the case of a trailing space on a comment line, because I've encountered this trouble today.

I'm editing my K8S configuration files with Intellij, and the default setting for this IDE is not complaining of having trailing spaces at the end of comment lines.

So when I've commited my latest modifications containing such a trailing whitespace, the github action in charge of validating my PR by calling Yamllint rejected it 😝

To me, it feels really awkward not been allowed to write whatever ones want on a comment line. 

So I guess it's more of a personal or dev team matter to allow it or not. 

That's why I think it should be configurable
